| **Inicio** | **atrás 12** | **Siguiente 14** |
|----------- |-------------- |---------------|
| [🏠](../../README.md) | [⏪](./12.DataFrames_con_Pandas.ipynb)| [⏩](./14.Scripts_Modulos.ipynb)|

# **13. Programación Orientada a Objetos en Python**

## **Clases**

En programación orientada a objetos `(POO)`, una clase es una estructura de datos que define un conjunto de atributos y métodos que se pueden utilizar para crear objetos. Los objetos son instancias de una clase y se pueden crear varias veces, cada uno con sus propios valores de atributos.

En Python, se define una clase mediante la palabra clave `class`, seguida del nombre de la clase, y después del contenido de la clase en un bloque indentado. Veamos un ejemplo:

In [1]:
class Perro:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad
    
    def ladrar(self):
        print("¡Guau!")

mi_perro = Perro("Fido", 3)
print(mi_perro.nombre)
mi_perro.ladrar()


Fido
¡Guau!


En este ejemplo, se define una clase `Perro` con dos atributos `(nombre y edad)` y un método `(ladrar)`. El método `__init__` es un método especial que se llama automáticamente cuando se crea un objeto de la clase y se utiliza para inicializar los atributos del objeto. En este caso, el método `__init__` toma dos argumentos (`nombre` y `edad`) y los asigna a los atributos correspondientes (`self.nombre` y `self.edad`).

Luego, se crea un objeto `mi_perro` de la clase `Perro` y se le asignan los valores `"Fido"` y `3` a sus atributos nombre y edad, respectivamente. Después, se imprime el valor del atributo nombre del objeto `mi_perro` y se llama al método ladrar del objeto, que imprime el mensaje `"¡Guau!"`.

Las clases pueden tener más de un método y pueden heredar atributos y métodos de otras clases. Por ejemplo:

In [3]:
class Animal:
    def __init__(self, especie):
        self.especie = especie

    def hacer_sonido(self):
        pass

class Perro(Animal):
    def __init__(self, nombre, edad):
        super().__init__("Perro")
        self.nombre = nombre
        self.edad = edad

    def hacer_sonido(self):
        print("¡Guau!")

mi_perro = Perro("Fido", 3)
print(mi_perro.especie)
mi_perro.hacer_sonido()


Perro
¡Guau!


En este ejemplo, se define una clase `Animal` con un atributo especie y un método `hacer_sonido`, que se define como una función vacía `(pass)` que se puede sobrescribir en clases hijas. Luego, se define una clase `Perro` que hereda de la clase `Animal` y que tiene su propio método `hacer_sonido` que imprime `"¡Guau!"`.

Cuando se crea un objeto `mi_perro` de la clase `Perro`, se llama al método `__init__` de la clase `Perro` que inicializa los atributos `nombre` y `edad` del objeto y también llama al método `__init__` de la clase `Animal` mediante la función `super()`, que inicializa el atributo especie del objeto con el valor `"Perro"`. Luego, se imprime el valor del atributo especie del objeto `mi_perro` y se llama al método `hacer_sonido` del objeto, que imprime `"¡Guau!"`.

## **Mi primera clase en Python**

En este ejemplo vamos a crear una clase simple para representar un círculo en un plano cartesiano. Para ello, utilizaremos los atributos `"x"` e `"y"` para representar las coordenadas del centro del círculo, y el atributo `"radio"` para representar su radio. Además, definiremos un método para calcular el área del círculo.

Aquí está el código de ejemplo:

In [4]:
class Circulo:
    def __init__(self, x, y, radio):
        self.x = x
        self.y = y
        self.radio = radio

    def area(self):
        return 3.14159 * self.radio ** 2


En este ejemplo, definimos una clase llamada `"Circulo"` que tiene tres atributos (`x`, `y` y `radio`) y un método (`area`). El método `__init__` es un método especial que se llama automáticamente cuando se crea un objeto de la clase y se utiliza para inicializar los atributos del objeto. En este caso, el método `__init__` toma tres argumentos (`x`, `y` y `radio`) y los asigna a los atributos correspondientes (`self.x`, `self.y` y `self.radio`).

El método `area` calcula y devuelve el área del círculo utilizando la fórmula `πr²`. En este caso, estamos utilizando una aproximación de `π` con un valor de `3.14159`.

Para utilizar esta clase, podemos crear un objeto de la clase `"Circulo"` y llamar a su método area. Por ejemplo:



In [5]:
mi_circulo = Circulo(0, 0, 5)
print(mi_circulo.area())


78.53975


En este caso, creamos un objeto `mi_circulo` de la clase `Circulo` con coordenadas `x=0`, `y=0` y `radio` `5`. Luego, llamamos al método area del objeto, que calcula y devuelve el área del círculo. En este caso, el resultado sería `78.53975`.

Esta es solo una muestra simple de cómo se puede utilizar una clase en Python para representar un objeto en un programa orientado a objetos. Las clases pueden tener muchos más atributos y métodos, y se pueden utilizar para representar objetos más complejos y sofisticados en una aplicación.

## **El método constructor**

El método constructor es un método especial en la programación orientada a objetos que se utiliza para inicializar los atributos de un objeto cuando se crea una instancia de una clase. En Python, el método constructor se llama `__init__()` y es el primer método que se ejecuta cuando se crea un objeto de una clase.

El método constructor toma como argumento el objeto que se está creando `(self)`, así como cualquier otro argumento que se desee inicializar en el objeto. Dentro del método constructor, los atributos del objeto se pueden inicializar utilizando los valores pasados como argumentos.

Aquí está un ejemplo de cómo se puede utilizar el método constructor en una clase de `perro`:

In [9]:
class Perro:
    def __init__(self, nombre, raza, edad):
        self.nombre = nombre
        self.raza = raza
        self.edad = edad


En este ejemplo, hemos definido una clase llamada `"Perro"` con tres atributos (`nombre`, `raza` y `edad`). El método `__init__` toma tres argumentos (`nombre`, `raza` y `edad`) y los asigna a los atributos correspondientes (`self.nombre`, `self.raza` y `self.edad`).

Para crear un objeto de la clase `"Perro"` con atributos específicos, se puede hacer lo siguiente:

In [15]:
mi_perro = Perro("Fido", "Labrador", 3)

En este caso, creamos un objeto `mi_perro` de la clase `Perro` con el nombre `"Fido"`, raza `"Labrador"` y edad `3`. El método constructor se llama automáticamente cuando se crea el objeto, y los atributos del objeto se inicializan con los valores pasados como argumentos.

Es importante tener en cuenta que el método constructor puede tener cualquier número de argumentos, y los argumentos pueden ser de cualquier tipo (como cadenas, números, listas, etc.). Además, la clase puede tener otros métodos además del constructor para realizar otras tareas relacionadas con el objeto.

## **Clase RationalNumber**

La clase `RationalNumber` en Python es una clase que representa un número racional, es decir, un número que puede ser expresado como una fracción con un numerador y un denominador. En esta clase, se pueden realizar operaciones aritméticas como la suma, la resta, la multiplicación y la división.

Aquí hay un ejemplo de cómo se puede definir la clase `RationalNumber`:

In [16]:
class RationalNumber:
    def __init__(self, numerador, denominador):
        self.numerador = numerador
        self.denominador = denominador

    def __add__(self, other):
        nuevo_denominador = self.denominador * other.denominador
        nuevo_numerador = (self.numerador * other.denominador) + (other.numerador * self.denominador)
        return RationalNumber(nuevo_numerador, nuevo_denominador)

    def __sub__(self, other):
        nuevo_denominador = self.denominador * other.denominador
        nuevo_numerador = (self.numerador * other.denominador) - (other.numerador * self.denominador)
        return RationalNumber(nuevo_numerador, nuevo_denominador)

    def __mul__(self, other):
        nuevo_numerador = self.numerador * other.numerador
        nuevo_denominador = self.denominador * other.denominador
        return RationalNumber(nuevo_numerador, nuevo_denominador)

    def __truediv__(self, other):
        nuevo_numerador = self.numerador * other.denominador
        nuevo_denominador = self.denominador * other.numerador
        return RationalNumber(nuevo_numerador, nuevo_denominador)

    def __str__(self):
        return "{}/{}".format(self.numerador, self.denominador)



En este ejemplo, la clase `RationalNumber` se define con un método constructor `__init__`, que toma un numerador y un denominador y los asigna a los atributos `self.numerador` y `self.denominador`. A continuación, se definen cuatro métodos especiales (`__add__`, `__sub__`, `__mul__` y `__truediv__`) que sobrecargan los operadores `+`, `-`, `*` y `/` para realizar operaciones aritméticas con objetos de tipo `RationalNumber`. Por último, se define el método `__str__`, que devuelve una representación en cadena del objeto `RationalNumber`.

Para utilizar la clase `RationalNumber`, se pueden crear objetos de tipo `RationalNumber` y realizar operaciones aritméticas con ellos. Por ejemplo:

In [17]:
r1 = RationalNumber(1, 2)
r2 = RationalNumber(3, 4)

r3 = r1 + r2
print(r3) # salida: 5/4

r4 = r1 * r2
print(r4) # salida: 3/8

r5 = r1 / r2
print(r5) # salida: 2/3


10/8
3/8
4/6


En este ejemplo, creamos dos objetos de tipo `RationalNumber` con valores de numerador y denominador, y luego realizamos operaciones aritméticas con ellos utilizando los métodos especiales sobrecargados. Cada operación aritmética devuelve un nuevo objeto `RationalNumber` que representa el resultado de la operación.

En resumen, la clase `RationalNumber` es una clase útil para representar números racionales en Python y realizar operaciones.

## **El método destructor**

El método destructor en Programación Orientada a Objetos en Python es un método especial que se utiliza para liberar recursos y limpiar objetos cuando ya no son necesarios. Este método se llama automáticamente cuando un objeto de una clase se elimina de la memoria o sale del alcance.

En Python, el método destructor se llama `__del__` y se define como cualquier otro método en la clase. El método `__del__` no toma ningún argumento además de `self`, que representa el objeto actual. En general, el método `__del__` se utiliza para liberar recursos externos, como cerrar archivos o conexiones de red, o para liberar memoria asignada dinámicamente.

Aquí hay un ejemplo de cómo se puede definir el método `__del__` en una clase:

In [18]:
class MyClass:
    def __init__(self, name):
        self.name = name
        print("Objeto creado: {}".format(self.name))

    def __del__(self):
        print("Objeto eliminado: {}".format(self.name))


En este ejemplo, se define la clase `MyClass` con un método constructor `__init__` que toma un nombre y lo asigna al atributo `self.name`. A continuación, se define el método destructor `__del__`, que simplemente imprime un mensaje que indica que el objeto se ha eliminado.

Para utilizar el método destructor, se pueden crear objetos de la clase `MyClass` y luego eliminarlos de la memoria. Por ejemplo:

In [19]:
obj1 = MyClass("objeto 1")
obj2 = MyClass("objeto 2")

del obj1
del obj2


Objeto creado: objeto 1
Objeto creado: objeto 2
Objeto eliminado: objeto 1
Objeto eliminado: objeto 2


En este ejemplo, se crean dos objetos de la clase `MyClass` y se eliminan de la memoria utilizando la instrucción `del`. Cada vez que se elimina un objeto, se llama automáticamente al método `__del__` de la clase correspondiente, que en este caso imprime un mensaje en la consola.

Es importante tener en cuenta que el método `__del__` no siempre se llama inmediatamente cuando se elimina un objeto. En lugar de eso, se llama en un momento indeterminado después de que el objeto se haya eliminado de la memoria. Por lo tanto, no se debe confiar en el método `__del__` para liberar recursos críticos o para garantizar un comportamiento específico del programa.

En resumen, el método destructor `__del__` en Programación Orientada a Objetos en Python es un método especial que se utiliza para liberar recursos y limpiar objetos cuando ya no son necesarios. Este método se llama automáticamente cuando un objeto se elimina de la memoria o sale del alcance.

## **Métodos de una clase**

En Programación Orientada a Objetos en Python, los métodos son funciones definidas dentro de una clase que se utilizan para realizar operaciones sobre los objetos creados a partir de esa clase. Los métodos definen el comportamiento de un objeto y permiten que el objeto realice acciones específicas y modifique su estado interno.

Existen dos tipos de métodos en Python: los métodos de instancia y los métodos de clase.

**Los métodos de instancia** son aquellos que se definen dentro de una clase y que operan sobre un objeto específico creado a partir de esa clase. Estos métodos utilizan el parámetro `self` para acceder a los atributos y métodos de la instancia.

Por ejemplo, supongamos que se tiene la siguiente clase `Persona`:

In [20]:
class Persona:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

    def saludar(self):
        print(f"Hola, mi nombre es {self.nombre} y tengo {self.edad} años.")


En este ejemplo, se define la clase `Persona` con un método constructor `__init__` que toma un nombre y una edad y los asigna a los atributos de la instancia. También se define el método `saludar`, que utiliza el atributo `nombre` y `edad` para imprimir un saludo personalizado.

Para utilizar estos métodos, se puede crear un objeto de la clase `Persona` y llamar al método `saludar`, como se muestra a continuación:

In [21]:
p = Persona("Juan", 30)
p.saludar()


Hola, mi nombre es Juan y tengo 30 años.


**Los métodos de clase**, por otro lado, se definen utilizando el decorador `@classmethod` y operan en la clase en sí misma en lugar de en una instancia específica. En estos métodos, el primer parámetro siempre es la propia clase `(cls)`.

Por ejemplo, supongamos que se tiene la siguiente clase `Empleado` con un método de clase `from_string`:

In [22]:
class Empleado:
    def __init__(self, nombre, salario):
        self.nombre = nombre
        self.salario = salario
    
    @classmethod
    def from_string(cls, string):
        nombre, salario = string.split("-")
        return cls(nombre, float(salario))


En este ejemplo, se define la clase `Empleado` con un método constructor `__init__` que toma un `nombre` y un `salario` y los asigna a los atributos de la instancia. También se define el método de clase `from_string`, que toma una cadena con el `nombre` y el `salario` separados por un guión y crea una nueva instancia de la clase `Empleado`.

Para utilizar este método, se puede llamar directamente a la clase `Empleado` y pasar la cadena como argumento:

In [23]:
e = Empleado.from_string("Juan-2500.50")
print(e.nombre)
print(e.salario)


Juan
2500.5


En resumen, los métodos en Programación Orientada a Objetos en Python son funciones definidas dentro de una clase que se utilizan para realizar operaciones sobre los objetos creados a partir de esa clase. Los métodos de instancia operan sobre un objeto específico y utilizan el parámetro `self`, mientras que los métodos de clase operan en la clase en sí misma y utilizan el parámetro `cls`.

## **Métodos de instancia**

En Programación Orientada a Objetos en Python, los métodos de instancia son aquellos que se definen dentro de una clase y operan sobre un objeto específico creado a partir de esa clase. Estos métodos utilizan el parámetro `self` para acceder a los atributos y métodos de la instancia.

Un método de instancia puede acceder a todos los atributos y métodos de la instancia y puede modificar su estado interno. Además, los métodos de instancia también pueden tomar argumentos adicionales, además del parámetro `self`.

Por ejemplo, considera la siguiente clase `Circulo`:

In [24]:
class Circulo:
    def __init__(self, radio):
        self.radio = radio
    
    def calcular_area(self):
        return 3.14 * (self.radio ** 2)


En este ejemplo, se define la clase `Circulo` con un método constructor `__init__` que toma un `radio` y lo asigna a un atributo de instancia llamado `radio`. También se define el método de instancia `calcular_area`, que utiliza el atributo `radio` para calcular y devolver el área del círculo.

Para utilizar estos métodos, se puede crear un objeto de la clase `Circulo` y llamar al método `calcular_area`, como se muestra a continuación:

In [25]:
c = Circulo(5)
print(c.calcular_area())


78.5


En este ejemplo, se crea un objeto `c` de la clase `Circulo` con un radio de `5` y se llama al método `calcular_area` para calcular y imprimir el área del círculo.

Además, los métodos de instancia también pueden tomar argumentos adicionales, además del parámetro `self`. Por ejemplo, se puede modificar la clase `Circulo` para que el método `calcular_area` tome un argumento adicional llamado `pi` que se utiliza en lugar del valor predeterminado de `3.14`:

In [26]:
class Circulo:
    def __init__(self, radio):
        self.radio = radio
    
    def calcular_area(self, pi=3.14):
        return pi * (self.radio ** 2)


En este ejemplo, el método `calcular_area` toma un argumento adicional llamado `pi` con un valor predeterminado de `3.14`. Si se proporciona un valor para `pi`, se utilizará ese valor en lugar del valor predeterminado de `3.14`.

Por ejemplo, se puede crear un objeto de la clase Circulo y llamar al método `calcular_area` con un valor personalizado para `pi`:

In [27]:
c = Circulo(5)
print(c.calcular_area(pi=3.14159265359))


78.53981633975


En resumen, los métodos de instancia en Programación Orientada a Objetos en Python son aquellos que se definen dentro de una clase y operan sobre un objeto específico creado a partir de esa clase. Estos métodos utilizan el parámetro `self` para acceder a los atributos y métodos de la instancia, y pueden tomar argumentos adicionales además del parámetro `self`.

## **Imprimiendo objetos RationalNumber en formato LaTeX**

En Programación Orientada a Objetos en Python, se puede imprimir un objeto `RationalNumber` en formato `LaTeX` utilizando el método especial `__repr__`. Este método debe devolver una cadena que representa al objeto de una manera legible y comprensible para los humanos.

Para imprimir un objeto `RationalNumber` en formato `LaTeX`, se puede utilizar la siguiente implementación del método `__repr__`:

In [28]:
class RationalNumber:
    def __init__(self, numerador, denominador):
        self.numerador = numerador
        self.denominador = denominador
    
    def __repr__(self):
        return "\\frac{" + str(self.numerador) + "}{" + str(self.denominador) + "}"


En este ejemplo, se define la clase `RationalNumber` con un método constructor `__init__` que toma un numerador y un denominador y los asigna a los atributos de instancia numerador y denominador, respectivamente. Además, se define el método especial `__repr__` que devuelve una cadena que representa al objeto en formato `LaTeX` utilizando la sintaxis de fracciones.

Para utilizar esta implementación del método `__repr__`, se puede crear un objeto `RationalNumber` y llamar a la función `print` para imprimir el objeto en formato `LaTeX`, como se muestra a continuación:

In [29]:
r = RationalNumber(2, 3)
print(r)


\frac{2}{3}


En este ejemplo, se crea un objeto `r` de la clase `RationalNumber` con un numerador de `2` y un denominador de `3`, y se llama a la función `print` para imprimir el objeto en formato `LaTeX`.

En resumen, se puede imprimir un objeto `RationalNumber` en formato `LaTeX` en Programación Orientada a Objetos en Python utilizando el método especial `__repr__` y la sintaxis de fracciones de `LaTeX`. Esto permite una representación más legible y comprensible de los objetos `RationalNumber` en los programas de Python.

## **Métodos estáticos**

En Programación Orientada a Objetos en Python, los métodos estáticos son métodos definidos dentro de una clase que no tienen acceso a los atributos de instancia de la clase y no necesitan una instancia de la clase para ser llamados. Estos métodos pueden ser llamados directamente desde la clase en sí, sin necesidad de crear una instancia de la misma.

Para definir un método estático en Python, se utiliza el decorador `@staticmethod`. La sintaxis básica es la siguiente:

```
class MyClass:
    @staticmethod
    def my_static_method(arg1, arg2):
        # código del método
```

En este ejemplo, se define un método estático `my_static_method` dentro de la clase `MyClass`. Este método no tiene acceso a los atributos de instancia de la clase y puede ser llamado directamente desde la clase utilizando la sintaxis `MyClass.my_static_method(arg1, arg2)`.

Un ejemplo de uso de un método estático es la creación de una función auxiliar que se utiliza en varias partes de un programa, pero que no está específicamente relacionada con ningún objeto o atributo de instancia en particular. Por ejemplo, se puede definir una clase `MathUtils` que contenga varios métodos estáticos para realizar cálculos matemáticos comunes:

In [31]:
class MathUtils:
    @staticmethod
    def factorial(n):
        if n < 0:
            raise ValueError("El factorial solo está definido para enteros no negativos.")
        elif n == 0 or n == 1:
            return 1
        else:
            return n * MathUtils.factorial(n-1)
    
    @staticmethod
    def fibonacci(n):
        if n < 0:
            raise ValueError("La serie de Fibonacci solo está definida para enteros no negativos.")
        elif n == 0:
            return 0
        elif n == 1:
            return 1
        else:
            return MathUtils.fibonacci(n-1) + MathUtils.fibonacci(n-2)


En este ejemplo, se define una clase `MathUtils` que contiene dos métodos estáticos, `factorial` y `fibonacci`. Estos métodos realizan cálculos matemáticos comunes y no requieren ningún atributo de instancia de la clase `MathUtils`. Por lo tanto, se pueden llamar directamente desde la clase utilizando la sintaxis `MathUtils.factorial(n)` o `MathUtils.fibonacci(n)`.

Para usar estos métodos en un programa, se puede llamar directamente a la clase `MathUtils` y sus métodos estáticos. Por ejemplo:

In [32]:
print(MathUtils.factorial(5))
print(MathUtils.fibonacci(7))


120
13


En resumen, los métodos estáticos en Programación Orientada a Objetos en Python son métodos definidos dentro de una clase que no tienen acceso a los atributos de instancia de la clase y no necesitan una instancia de la clase para ser llamados. Estos métodos son útiles para crear funciones auxiliares que no están específicamente relacionadas con ningún objeto en particular, pero que se utilizan en varias partes de un programa.

## **Métodos estáticos de la clase RationalNumber**

En Programación Orientada a Objetos en Python, los métodos estáticos de la clase `RationalNumber` son métodos definidos dentro de la clase que no tienen acceso a los atributos de instancia de la misma y no requieren una instancia de la clase para ser llamados. Estos métodos pueden ser útiles para realizar cálculos matemáticos comunes que no están específicamente relacionados con ningún objeto `RationalNumber` en particular.

Para definir un método estático en Python, se utiliza el decorador `@staticmethod`. La sintaxis básica es la siguiente:

```
class MyClass:
    @staticmethod
    def my_static_method(arg1, arg2):
        # código del método
```

En el caso de la clase `RationalNumber`, se podría definir un método estático para calcular el máximo común divisor `(MCD)` de dos números enteros utilizando el algoritmo de Euclides. El código podría verse así:

In [33]:
class RationalNumber:
    # código para el constructor, métodos de instancia, etc.
    
    @staticmethod
    def gcd(a, b):
        """
        Calcula el máximo común divisor (MCD) de dos números enteros utilizando el algoritmo de Euclides.
        """
        while b:
            a, b = b, a % b
        return a


En este ejemplo, se define un método estático `gcd` dentro de la clase `RationalNumber`. Este método utiliza el algoritmo de Euclides para calcular el máximo común divisor de dos números enteros `a` y `b`. Como este método no tiene acceso a los atributos de instancia de la clase `RationalNumber`, se puede definir como un método estático.

Para usar este método en un programa, se puede llamar directamente a la clase `RationalNumber` y su método estático. Por ejemplo, para calcular el `MCD` de `12` y `18`, se podría hacer lo siguiente:



In [34]:
print(RationalNumber.gcd(12, 18))  # salida: 6


6


En este ejemplo, se llama al método estático `gcd` de la clase `RationalNumber` pasando los valores `12` y `18` como argumentos. El método calcula el `MCD` de estos dos números utilizando el algoritmo de Euclides y devuelve el resultado, que se imprime en la consola.

En resumen, los métodos estáticos de la clase `RationalNumber` en Programación Orientada a Objetos en Python son métodos definidos dentro de la clase que no tienen acceso a los atributos de instancia de la misma y no requieren una instancia de la clase para ser llamados. Estos métodos pueden ser útiles para realizar cálculos matemáticos comunes que no están específicamente relacionados con ningún objeto `RationalNumber` en particular.

## **Métodos de clase**

En Programación Orientada a Objetos en Python, los métodos de clase son métodos que operan en la clase en sí misma en lugar de en una instancia de la clase. Es decir, los métodos de clase no tienen acceso a los atributos de instancia de la clase y se utilizan para realizar operaciones que son relevantes para la clase en su conjunto, en lugar de para una instancia específica.

Para definir un método de clase en Python, se utiliza el decorador `@classmethod`. La sintaxis básica es la siguiente:

```
class MyClass:
    @classmethod
    def my_class_method(cls, arg1, arg2):
        # código del método
```

En el caso de la clase `RationalNumber`, se podría definir un método de clase para crear un nuevo objeto `RationalNumber` a partir de una cadena en formato de fracción. El código podría verse así:

Los métodos de clase en Programación Orientada a Objetos en Python son aquellos que están asociados con la clase en sí misma, en lugar de con una instancia particular de la clase. A diferencia de los métodos de instancia, los métodos de clase no reciben como primer parámetro la referencia al objeto `(self)`, sino que reciben la referencia a la propia clase `(cls)`.

Los métodos de clase se definen usando el decorador `@classmethod`. Por lo general, se utilizan para crear objetos de una clase de formas distintas a las habituales, o para realizar operaciones que involucren a la clase en su conjunto, en lugar de a una instancia particular.

A continuación, se muestra un ejemplo de cómo definir y utilizar un método de clase en Python:

In [45]:
class MyClass:
    my_var = 42

    @classmethod
    def from_string(cls, s):
        """
        Crea una instancia de la clase MyClass a partir de una cadena de texto.
        """
        return cls(int(s))

    def __init__(self, x):
        self.x = x

    def my_method(self):
        print("El valor de x es:", self.x)

# Crear una instancia de MyClass utilizando el método from_string
c = MyClass.from_string("10")
c.my_method() # Imprime "El valor de x es: 10"


El valor de x es: 10


En este ejemplo, se define una clase `MyClass` que tiene un atributo de clase `my_var`, un método de instancia `__init__` y otro método de instancia `my_method`. Además, se define un método de clase `from_string`, que toma una cadena de texto y crea una instancia de la clase `MyClass` a partir de ella. La diferencia con respecto a un método de instancia es que `from_string` recibe como primer parámetro la referencia a la clase `(cls)`, en lugar de la referencia al objeto `(self)`.

Para utilizar el método `from_string`, se llama directamente a la clase `MyClass` y se invoca el método utilizando la sintaxis de punto. El resultado es una instancia de `MyClass` cuyo atributo `x` tiene el valor `10`.






## **Métodos de clase de la clase RationalNumber**

En programación orientada a objetos en Python, los métodos de clase son aquellos que se definen a nivel de la clase y no a nivel de la instancia. Es decir, estos métodos no operan sobre los atributos específicos de una instancia de la clase, sino que se utilizan para realizar operaciones en la propia clase.

En la clase `RationalNumber` que hemos venido desarrollando, podemos definir métodos de clase que nos permitan realizar operaciones que involucren a todas las instancias de la clase `RationalNumber`, y no solo a una instancia específica.

Para definir un método de clase en Python, se utiliza el decorador `@classmethod`. Este decorador se aplica a un método que se va a definir dentro de la clase y le indica a Python que ese método es un método de clase en lugar de un método de instancia.

A continuación, un ejemplo de cómo se puede definir un método de clase en la clase `RationalNumber`:

In [52]:
class RationalNumber:
    def __init__(self, numerator, denominator=1):
        self.numerator = numerator
        self.denominator = denominator

    @classmethod
    def from_string(cls, s):
        """
        Crea una instancia de la clase RationalNumber a partir de una cadena de texto que represente una fracción.
        """
        numerator, denominator = map(int, s.split("/"))
        return cls(numerator, denominator)

    @classmethod
    def gcd(cls, a, b):
        """
        Calcula el máximo común divisor entre dos números.
        """
        while b:
            a, b = b, a % b
        return a

    # rest of the code


En este ejemplo, se han definido dos métodos de clase:

* **from_string(cls, s):**

 Este método permite crear una instancia de la clase `RationalNumber` a partir de una cadena de texto que represente una fracción. El método se define con el decorador `@classmethod`, lo que indica que es un método de clase. Al igual que con los métodos de instancia, el primer parámetro de un método de clase es `cls`, que hace referencia a la clase misma.

* **gcd(cls, a, b):**

 Este método calcula el máximo común divisor entre dos números. Nuevamente, el método se define con el decorador `@classmethod` y recibe como primer parámetro la propia clase `(cls)`.

Para utilizar estos métodos de clase, se llama al método directamente desde la clase, sin necesidad de crear una instancia previamente. Por ejemplo:

In [53]:
r1 = RationalNumber(3, 4)
r2 = RationalNumber.from_string("5/6")

print(RationalNumber.gcd(r1.numerator, r1.denominator))  # Output: 1
print(RationalNumber.gcd(r2.numerator, r2.denominator))  # Output: 5


1
1


En este ejemplo, se ha utilizado el método `from_string` para crear una nueva instancia de la clase `RationalNumber` a partir de una cadena de texto. También se ha utilizado el método `gcd` para calcular el máximo común divisor entre el numerador y el denominador de dos instancias de la clase `RationalNumber` `(r1 y r2)`. En ambos casos, se ha llamado al método directamente desde la clase, sin necesidad de crear una instancia previamente.

En resumen, los métodos de clase en Python permiten definir métodos que operan sobre la propia clase y no sobre una instancia específica de la clase. Estos métodos se definen con el decorador `@classmethod`

## **Propiedades**

Las propiedades, en programación orientada a objetos en Python, son una forma de definir atributos de una clase que se comportan como variables públicas, pero que permiten controlar el acceso a sus valores mediante los métodos `getter` y `setter`. De esta forma, se pueden establecer restricciones sobre el valor que se asigna a un atributo, o bien, realizar alguna operación específica antes o después de leer o modificar el valor de dicho atributo.

Para definir una propiedad en Python, se utiliza el decorador `@property` antes del método que se desea que funcione como `getter`, y se utiliza el decorador `@<nombre_atributo>`.`setter` antes del método que se desea que funcione como `setter`, donde `<nombre_atributo>` es el nombre del atributo que se quiere controlar.

Veamos un ejemplo para entenderlo mejor:

In [54]:
class Persona:
    def __init__(self, nombre, edad):
        self._nombre = nombre
        self._edad = edad

    @property
    def nombre(self):
        return self._nombre

    @nombre.setter
    def nombre(self, nuevo_nombre):
        if not isinstance(nuevo_nombre, str):
            raise ValueError("El nombre debe ser una cadena de texto")
        self._nombre = nuevo_nombre

    @property
    def edad(self):
        return self._edad

    @edad.setter
    def edad(self, nueva_edad):
        if not isinstance(nueva_edad, int):
            raise ValueError("La edad debe ser un número entero")
        if nueva_edad < 0:
            raise ValueError("La edad no puede ser negativa")
        self._edad = nueva_edad


En este ejemplo, definimos la clase Persona con dos atributos: nombre y edad. Para controlar el acceso a estos atributos, definimos dos métodos `getter (@property)` y dos métodos `setter` `(@nombre.setter y @edad.setter)`. Estos métodos nos permiten acceder y modificar el valor de los atributos de la clase, pero al mismo tiempo, nos permiten definir algunas restricciones sobre el valor que se asigna.

Por ejemplo, en el `setter` del atributo nombre, validamos que el valor que se está asignando sea una cadena de texto. Si no lo es, lanzamos una excepción. De esta forma, nos aseguramos de que siempre se esté asignando un valor válido a este atributo.

Del mismo modo, en el `setter` del atributo edad, validamos que el valor que se está asignando sea un número entero y que no sea negativo. De esta forma, nos aseguramos de que siempre se esté asignando un valor válido a este atributo.

Para utilizar esta clase, podemos crear una instancia de la misma y acceder a sus atributos como si fueran variables públicas:

In [55]:
p = Persona("Juan", 25)
print(p.nombre) # Juan
print(p.edad) # 25

p.nombre = "Pedro"
p.edad = 30

print(p.nombre) # Pedro
print(p.edad) # 30


Juan
25
Pedro
30


En este ejemplo, creamos una instancia de la clase `Persona` y le asignamos los valores "Juan" y 25 a los atributos nombre y edad, respectivamente. Después, accedemos a estos atributos utilizando los métodos `getter` `(@property)`. Finalmente, modificamos el valor de estos atributos utilizando los métodos `setter` `(@nombre.setter y @edad.setter)` y volvemos a acceder a ellos para comprobar que han cambiado.

## **Propiedades de la clase RationalNumber**

Las propiedades en programación orientada a objetos en Python son una forma de acceder y modificar atributos de una clase como si fueran atributos normales, pero en realidad son métodos que se definen con un decorador especial `@property`.

La clase `RationalNumber` ya tiene definidos los métodos numerator y denominator que devuelven los atributos `_numerator` y `_denominator`, respectivamente. Ahora vamos a convertir estos métodos en propiedades.

Para ello, agregamos el decorador `@property` antes de la definición de cada método y eliminamos los parámetros de entrada.

In [56]:
class RationalNumber:
    def __init__(self, numerator, denominator):
        self._numerator = numerator
        self._denominator = denominator

    @property
    def numerator(self):
        return self._numerator

    @property
    def denominator(self):
        return self._denominator

    def __repr__(self):
        return f"{self._numerator}/{self._denominator}"


Ahora podemos acceder a estos métodos como si fueran atributos normales:

In [57]:
r = RationalNumber(3, 4)
print(r.numerator)   # Output: 3
print(r.denominator) # Output: 4


3
4


También podemos modificar los valores de estos atributos a través de métodos especiales llamados `"setter"`, que se definen utilizando el decorador `@<property>.setter`. Por ejemplo, para agregar un método que permita cambiar el numerador de la fracción:



In [58]:
class RationalNumber:
    def __init__(self, numerator, denominator):
        self._numerator = numerator
        self._denominator = denominator

    @property
    def numerator(self):
        return self._numerator

    @numerator.setter
    def numerator(self, value):
        self._numerator = value

    @property
    def denominator(self):
        return self._denominator

    def __repr__(self):
        return f"{self._numerator}/{self._denominator}"


Ahora podemos modificar el numerador de la fracción:

In [59]:
r = RationalNumber(3, 4)
print(r.numerator)   # Output: 3
r.numerator = 5
print(r.numerator)   # Output: 5


3
5


## **Herencia de clases**

La herencia de clases es un concepto fundamental en la Programación Orientada a Objetos `(POO)` que permite a una clase hija (o subclase) heredar propiedades y métodos de su clase padre (o superclase). Esto significa que la subclase tiene acceso a todas las propiedades y métodos públicos y protegidos de su superclase, y puede agregar o sobrescribir sus propios métodos o propiedades.

En Python, para definir una `subclase`, simplemente se especifica la `superclase` entre paréntesis después del nombre de la `subclase`. Por ejemplo:

In [60]:
class Animal:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def eat(self):
        print(f"{self.name} is eating.")

class Dog(Animal):
    def bark(self):
        print("Woof!")

dog = Dog("Rex", 3)
print(dog.name)
dog.eat()
dog.bark()


Rex
Rex is eating.
Woof!


En este ejemplo, la clase `Animal` es la `superclase` y la clase `Dog` es la `subclase`. La clase `Dog` hereda la propiedad `name` y el método `eat()` de la clase `Animal`, y agrega su propio método `bark()`. El objeto `dog` es una instancia de la clase `Dog` y puede acceder a las propiedades y métodos de ambas clases.

Cabe destacar que si un método o propiedad se define en la `subclase` con el mismo nombre que uno definido en la `superclase`, entonces la versión de la `subclase` sobrescribe la versión de la `superclase`. Además, las `subclases` pueden tener sus propias `subclases`, lo que permite crear una jerarquía de clases.

La herencia de clases es útil para evitar la duplicación de código y promover la reutilización de código. También permite una mayor flexibilidad en el diseño de programas, ya que las `subclases` pueden adaptar y especializar el comportamiento de las `superclases` para satisfacer sus necesidades específicas.

En resumen, la herencia de clases es un concepto clave en la Programación Orientada a Objetos que permite la creación de jerarquías de clases y la reutilización de código. En Python, la herencia se logra especificando la superclase entre paréntesis después del nombre de la subclase.

## **Herencia Simple**

La herencia simple es un concepto de programación orientada a objetos que permite crear una nueva clase basada en una clase existente (llamada clase base o clase padre). La nueva clase (llamada clase derivada o clase hija) hereda todos los atributos y métodos de la clase base y puede agregar o redefinir sus propios atributos y métodos.

En Python, la herencia simple se implementa mediante la declaración de la clase hija con la clase padre como argumento entre paréntesis en la definición de la clase. La sintaxis básica es la siguiente:

```
class ClasePadre:
    # definición de la clase padre

class ClaseHija(ClasePadre):
    # definición de la clase hija
```

En este ejemplo, `ClaseHija` es una subclase de `ClasePadre`. La clase hija hereda todos los métodos y atributos de la clase padre y puede definir sus propios métodos y atributos. Además, la clase hija puede sobrescribir los métodos de la clase padre para cambiar su comportamiento.

Veamos un ejemplo de la herencia simple en Python:

In [61]:
class Animal:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

    def hablar(self):
        print("El animal está hablando.")

class Perro(Animal):
    def __init__(self, nombre, edad, raza):
        super().__init__(nombre, edad)
        self.raza = raza

    def hablar(self):
        print("El perro está ladrando.")

    def correr(self):
        print("El perro está corriendo.")


En este ejemplo, `Animal` es la clase base y `Perro` es la clase derivada. La clase `Perro` hereda los atributos `nombre` y `edad` de la clase `Animal`, y define su propio atributo `raza`. La clase `Perro` también sobrescribe el método hablar de la clase `Animal` para que el perro ladre en lugar de hablar genéricamente. Además, la clase `Perro` define un nuevo método correr.

Podemos crear instancias de ambas clases y llamar a sus métodos de la siguiente manera:

In [62]:
animal = Animal("Bambi", 2)
perro = Perro("Fido", 3, "Labrador")

print(animal.nombre)
animal.hablar()

print(perro.nombre)
print(perro.raza)
perro.hablar()
perro.correr()


Bambi
El animal está hablando.
Fido
Labrador
El perro está ladrando.
El perro está corriendo.


Podemos ver cómo la clase `Perro` hereda los atributos y métodos de la clase `Animal`, y cómo la clase `Perro` redefine y extiende algunos de estos atributos y métodos. La herencia simple es una herramienta poderosa para reutilizar el código y crear jerarquías de clases bien organizadas y estructuradas.

## **Sobrescribiendo métodos**

En programación orientada a objetos, la sobrescritura de métodos es una técnica que permite a una `subclase` proporcionar su propia implementación de un método ya definido en su clase base. Esta técnica es muy útil cuando queremos extender la funcionalidad de una clase existente sin modificar su comportamiento original.

En Python, la sobrescritura de métodos se logra simplemente definiendo un método con el mismo nombre en la `subclase` que el método que se desea sobrescribir en la clase base. Cuando se llama al método en la instancia de la `subclase`, la implementación de la `subclase` se utilizará en lugar de la implementación de la clase base.

Veamos un ejemplo para entender mejor la sobrescritura de métodos en Python:

In [63]:
class Animal:
    def speak(self):
        print("El animal hace un sonido genérico")

class Perro(Animal):
    def speak(self):
        print("Guau!")

class Gato(Animal):
    def speak(self):
        print("Miau!")


En este ejemplo, definimos una clase `Animal` con un método `speak()` que imprime un mensaje genérico. Luego, definimos dos subclases, `Perro` y `Gato`, que heredan de `Animal`. Cada subclase sobrescribe el método `speak()` con su propia implementación que representa el sonido que hace ese tipo de animal.

Ahora, si creamos instancias de `Perro` y `Gato` y llamamos a su método `speak()`, veremos que cada instancia produce un sonido diferente:

In [64]:
mi_perro = Perro()
mi_gato = Gato()

mi_perro.speak()    # Output: "Guau!"
mi_gato.speak()     # Output: "Miau!"


Guau!
Miau!


Aquí podemos ver que la sobrescritura de métodos nos permitió personalizar el comportamiento de las subclases sin tener que modificar la clase base `Animal`. En su lugar, simplemente proporcionamos una implementación diferente del método `speak()` en cada subclase.

## **El método .super() en single inheritance**

En Python, la función `super()` se utiliza para acceder a los métodos de la clase padre desde una clase hija. Esto es especialmente útil cuando estamos utilizando herencia, ya que podemos extender las funcionalidades de la clase padre y, al mismo tiempo, mantener sus comportamientos originales.

Cuando se utiliza `super()` en una clase hija, se refiere a la clase padre de la cual se está heredando. De esta manera, podemos llamar a los métodos de la clase padre y utilizar sus argumentos en la clase hija.

Veamos un ejemplo sencillo. Supongamos que tenemos una clase padre llamada Animal con un método llamado `hacer_sonido`:

In [68]:
class Animal:
    def hacer_sonido(self):
        print('El animal está haciendo un sonido.')


Luego, creamos una clase hija llamada `Perro`, que hereda de la clase padre `Animal`. En esta clase, queremos sobrescribir el método `hacer_sonido` para que el perro ladre en lugar de hacer un sonido genérico:

In [69]:
class Perro(Animal):
    def hacer_sonido(self):
        print('El perro está ladrando.')


Ahora, si creamos una instancia de la clase `Perro` y llamamos al método `hacer_sonido`

In [70]:
perro = Perro()
perro.hacer_sonido()


El perro está ladrando.


Pero, ¿cómo podemos acceder al método `hacer_sonido` de la clase padre desde la clase hija? Es aquí donde entra en juego el método `super()`. Podemos utilizar `super()` en la clase hija para llamar al método de la clase padre:

In [71]:
class Perro(Animal):
    def hacer_sonido(self):
        super().hacer_sonido()
        print('El perro está ladrando.')


En este ejemplo, `super().hacer_sonido()` llama al método `hacer_sonido` de la clase padre `Animal`, y luego imprimimos "El perro está ladrando." en una nueva línea. Ahora, si creamos una instancia de la clase `Perro` y llamamos al método hacer_sonido, obtendremos el siguiente resultado:

In [72]:
perro = Perro()
perro.hacer_sonido()

El animal está haciendo un sonido.
El perro está ladrando.


Observa que ahora también se imprime "El animal está haciendo un sonido." antes de "El perro está ladrando.", ya que estamos llamando al método `hacer_sonido` de la clase padre con `super().hacer_sonido()`.

## **Herencia Múltiple**

La herencia múltiple en programación orientada a objetos en Python es un mecanismo que permite a una clase heredar atributos y métodos de varias clases padres. En Python, una clase puede heredar de múltiples clases y así obtener todas las características de ellas.

La herencia múltiple se utiliza cuando una clase requiere heredar comportamientos y atributos de varias clases. Es importante destacar que, aunque puede ser muy útil, también puede generar problemas de ambigüedad si una clase padre define un atributo o método con el mismo nombre que otra clase padre.

En Python, la herencia múltiple se logra al especificar varias clases entre paréntesis después del nombre de la clase. Cuando se hereda de varias clases, la clase hija hereda todos los atributos y métodos de sus padres.

A continuación, se muestra un ejemplo de cómo se usa la herencia múltiple en Python:

In [74]:
class A:
    def method_a(self):
        print("Este es el método A")

class B:
    def method_b(self):
        print("Este es el método B")
        
class C(A, B):
    pass

c = C()
c.method_a() # Output: "Este es el método A"
c.method_b() # Output: "Este es el método B"


Este es el método A
Este es el método B


En este ejemplo, tenemos tres clases: `A`, `B` y `C`. La clase `C` hereda de ambas clases `A` y `B` utilizando la herencia múltiple. La clase `C` no tiene ningún método propio, por lo que simplemente pasamos la instrucción `pass`. La instancia de la clase `C` tiene acceso a los métodos `method_a` y `method_b` de las clases `A` y `B` respectivamente.

## **El método .super() en múltiple inheritance**

En Python, el método `super()` se utiliza en la herencia de clases para acceder a los métodos de una clase padre desde una subclase. En el caso de la herencia múltiple, `super()` se utiliza para acceder a los métodos de varias clases padre en el orden en que se especifican.

La sintaxis general del método `super()` es la siguiente:

`super().nombre_del_método(args)`

donde `nombre_del_método` es el nombre del método que se desea llamar y `args` son los argumentos que se pasan al método. Es importante destacar que no se especifica el nombre de la clase padre ya que `super()` se encarga automáticamente de encontrar la siguiente clase en la jerarquía de herencia.

Veamos un ejemplo de cómo usar `super()` en la herencia múltiple:

In [78]:
class A:
    def __init__(self):
        print("Inicializando clase A")
        
    def metodo(self):
        print("Método de la clase A")

class B:
    def __init__(self):
        print("Inicializando clase B")
        
    def metodo(self):
        print("Método de la clase B")

class C(A, B):
    def __init__(self):
        print("Inicializando clase C")
        super().__init__()

    def metodo(self):
        print("Método de la clase C")
        super().metodo()

c = C()


Inicializando clase C
Inicializando clase A


En este ejemplo, `C` es una subclase que hereda de `A` y `B`. En el método `__init__` de la clase `C`, usamos `super()` para llamar al método `__init__` de la clase `A` en primer lugar. Luego, en el método metodo de `C`, usamos `super()` para llamar al método metodo de la clase `A`, y luego al método metodo de la clase `B`. En este caso, el orden en que se especifican las clases padre en la definición de la clase `C` es importante porque determina el orden en que se llaman los métodos.

En resumen, el método `super()` es una forma útil de acceder a los métodos de una clase padre en la herencia de clases en Python, especialmente en el caso de la herencia múltiple donde se debe especificar el orden en que se llaman los métodos de las clases padre.

## **Polimorfismo**

El polimorfismo es uno de los conceptos fundamentales en la Programación Orientada a Objetos `(POO)` que permite que los objetos de diferentes clases se comporten de manera similar. El polimorfismo permite que un objeto pueda ser tratado como otro objeto, aunque sean de clases distintas, siempre y cuando tengan métodos o atributos en común.

El polimorfismo se puede lograr de diferentes maneras en Python, una de ellas es mediante la definición de métodos con el mismo nombre en diferentes clases, pero con implementaciones diferentes. Esto se conoce como polimorfismo de sobrecarga. Otra forma es mediante el uso de la herencia y la sobrescritura de métodos, lo que se conoce como polimorfismo de anulación.

Veamos un ejemplo de polimorfismo de sobrecarga en Python:

In [79]:
class Figura:
    def area(self):
        pass

class Rectangulo(Figura):
    def __init__(self, base, altura):
        self.base = base
        self.altura = altura
    
    def area(self):
        return self.base * self.altura

class Circulo(Figura):
    def __init__(self, radio):
        self.radio = radio
    
    def area(self):
        return 3.14 * self.radio ** 2


En este ejemplo, creamos tres clases: `Figura`, `Rectangulo` y `Circulo`. La clase Figura es una clase base abstracta que define un método `area()` que no tiene implementación. Las clases `Rectangulo` y `Circulo` son subclases de `Figura` y sobrescriben el método `area()` con su propia implementación para calcular el área de un rectángulo o un círculo.

Podemos crear objetos de ambas subclases y tratarlos como objetos de la clase base Figura, ya que ambos tienen un método `area()` definido:

In [80]:
figuras = [Rectangulo(3, 4), Circulo(5)]
for figura in figuras:
    print(figura.area())


12
78.5


Este código crea una lista de objetos `Rectangulo` y `Circulo` y los recorre usando un bucle `for`. En cada iteración, se llama al método `area()` de la figura actual, que puede ser un objeto de cualquiera de las dos clases. Esto se debe a que ambos objetos tienen un método `area()` con el mismo nombre, y Python utiliza el polimorfismo para llamar al método correcto en cada caso.

En resumen, el polimorfismo es una técnica que permite que los objetos de diferentes clases se comporten de manera similar, lo que hace que el código sea más flexible y fácil de mantener. En Python, el polimorfismo se puede lograr de diferentes maneras, como la sobrecarga de métodos y la herencia de clases.






## **Variables privadas en Python: la técnica del Mangling**

En Python, no existe una verdadera encapsulación de datos en el sentido tradicional, como lo hay en otros lenguajes de programación orientados a objetos. Sin embargo, Python proporciona una técnica llamada `"mangling"` para simular la privacidad de las variables de instancia y los métodos de una clase.

La técnica del mangling consiste en agregar un doble guion bajo al inicio del nombre de la variable o método, lo que hace que Python la trate como privada y la haga más difícil de acceder desde fuera de la clase.

Veamos un ejemplo de cómo se usa la técnica de mangling en Python:

In [82]:
class Persona:
    def __init__(self, nombre, edad):
        self.__nombre = nombre
        self.__edad = edad
    
    def get_edad(self):
        return self.__edad
    
    def set_edad(self, edad):
        if edad > 0:
            self.__edad = edad
    
    def __get_nombre(self):
        return self.__nombre
    
    def __set_nombre(self, nombre):
        if nombre.isalpha():
            self.__nombre = nombre
    
    nombre = property(fget=__get_nombre, fset=__set_nombre)


En este ejemplo, hemos definido una clase `Persona` con dos variables de instancia privadas `__nombre` y `__edad`. También hemos definido dos métodos `get_edad()` y `set_edad()` para acceder y modificar la variable de instancia `__edad`, respectivamente.

Además, hemos definido dos métodos privados `__get_nombre()` y `__set_nombre()` que se utilizarán para acceder y modificar la variable de instancia `__nombre`. Sin embargo, en lugar de llamar directamente a estos métodos privados, hemos creado una propiedad llamada nombre que se utilizará para acceder y modificar la variable de instancia `__nombre`.

La técnica de mangling hace que estas variables y métodos sean más difíciles de acceder desde fuera de la clase, aunque aún es posible hacerlo utilizando una sintaxis especial. Por ejemplo, para acceder a la variable de instancia privada `__nombre`, podemos hacer lo siguiente:

In [83]:
p = Persona("Juan", 25)
print(p._Persona__nombre)


Juan


Aquí, estamos accediendo directamente a la variable de instancia privada `__nombre` utilizando la sintaxis de mangling `_Persona__nombre`. Sin embargo, esta es una práctica no recomendada, ya que viola la privacidad de los datos de la clase.

En resumen, la técnica de mangling en Python permite simular la privacidad de las variables y métodos de una clase, agregando un doble guion bajo al inicio de su nombre. Sin embargo, esta técnica no proporciona una verdadera encapsulación de datos y aún es posible acceder a estas variables y métodos desde fuera de la clase utilizando una sintaxis especial.

## **Clases y Objetos**

Python es un lenguaje de programación orientado a objetos. Todo en Python es un objeto, con sus propiedades y métodos. Un número, cadena, lista, diccionario, tupla, conjunto, etc. utilizado en un programa es un objeto de una clase integrada correspondiente. Creamos clase para crear un objeto. Una clase es como un constructor de objetos o un "modelo" para crear objetos. Instanciamos una clase para crear un objeto. La clase define los atributos y el comportamiento del objeto, mientras que el objeto, por otro lado, representa la clase.

Hemos estado trabajando con clases y objetos desde el comienzo de este desafío sin saberlo. Cada elemento en un programa de Python es un objeto de una clase. Comprobemos si todo en python es una clase:

### **Crear una clase**

Para crear una clase necesitamos la palabra clave `clase` seguida del nombre y dos puntos. El nombre de la clase debe ser ```CamelCase```.

```
# syntax
class ClassName:
  code goes here
```

In [84]:
class Person:
  pass
print(Person)

<class '__main__.Person'>


### **Crear un objeto**

Podemos crear un objeto llamando a la `clase`.

In [85]:
p = Person()
print(p)

### **Constructor de clase**

En los ejemplos anteriores, hemos creado un objeto de la clase `Person`. Sin embargo, una clase sin constructor no es realmente útil en aplicaciones reales. Usemos la función constructora para que nuestra clase sea más útil. Al igual que la función constructora en Java o JavaScript, Python también tiene una función constructora `init()` incorporada . La función constructora `init` tiene un parámetro propio que es una referencia a la instancia actual de la clase.

**Ejemplos:**

In [86]:
class Person:
      def __init__ (self, name):
        # self allows to attach parameter to the class
          self.name =name

p = Person('Asabeneh')
print(p.name)
print(p)

Asabeneh


Agreguemos más parámetros a la función constructora.

In [87]:
class Person:
      def __init__(self, firstname, lastname, age, country, city):
          self.firstname = firstname
          self.lastname = lastname
          self.age = age
          self.country = country
          self.city = city


p = Person('Asabeneh', 'Yetayeh', 25, 'Finland', 'Helsinki')
print(p.firstname)
print(p.lastname)
print(p.age)
print(p.country)
print(p.city)

Asabeneh
Yetayeh
25
Finland
Helsinki


### **Métodos de objetos**

Los objetos pueden tener métodos. Los métodos son funciones que pertenecen al objeto.

In [88]:
class Person:
      def __init__(self, firstname, lastname, age, country, city):
          self.firstname = firstname
          self.lastname = lastname
          self.age = age
          self.country = country
          self.city = city
      def person_info(self):
        return f'{self.firstname} {self.lastname} is {self.age} years old. He lives in {self.city}, {self.country}'

p = Person('Asabeneh', 'Yetayeh', 250, 'Finland', 'Helsinki')
print(p.person_info())

Asabeneh Yetayeh is 250 years old. He lives in Helsinki, Finland


### **Métodos predeterminados de objetos**

A veces, es posible que desee tener valores predeterminados para sus métodos de objeto. Si damos valores predeterminados para los parámetros en el constructor, podemos evitar errores cuando llamamos o instanciamos nuestra clase sin parámetros. Veamos cómo se ve:

In [89]:
class Person:
      def __init__(self, firstname='Asabeneh', lastname='Yetayeh', age=250, country='Finland', city='Helsinki'):
          self.firstname = firstname
          self.lastname = lastname
          self.age = age
          self.country = country
          self.city = city

      def person_info(self):
        return f'{self.firstname} {self.lastname} is {self.age} years old. He lives in {self.city}, {self.country}.'

p1 = Person()
print(p1.person_info())
p2 = Person('John', 'Doe', 30, 'Nomanland', 'Noman city')
print(p2.person_info())

Asabeneh Yetayeh is 250 years old. He lives in Helsinki, Finland.
John Doe is 30 years old. He lives in Noman city, Nomanland.


### **Método para modificar los valores predeterminados de la clase**

En el siguiente ejemplo, la clase de `persona`, todos los parámetros del constructor tienen valores predeterminados. Además de eso, tenemos el parámetro de habilidades, al que podemos acceder usando un método. Vamos a crear el método `add_skill` para agregar habilidades a la lista de habilidades.

In [90]:
class Person:
      def __init__(self, firstname='Asabeneh', lastname='Yetayeh', age=250, country='Finland', city='Helsinki'):
          self.firstname = firstname
          self.lastname = lastname
          self.age = age
          self.country = country
          self.city = city
          self.skills = []

      def person_info(self):
        return f'{self.firstname} {self.lastname} is {self.age} years old. He lives in {self.city}, {self.country}.'
      def add_skill(self, skill):
          self.skills.append(skill)

p1 = Person()
print(p1.person_info())
p1.add_skill('HTML')
p1.add_skill('CSS')
p1.add_skill('JavaScript')
p2 = Person('John', 'Doe', 30, 'Nomanland', 'Noman city')
print(p2.person_info())
print(p1.skills)
print(p2.skills)

Asabeneh Yetayeh is 250 years old. He lives in Helsinki, Finland.
John Doe is 30 years old. He lives in Noman city, Nomanland.
['HTML', 'CSS', 'JavaScript']
[]


## **Herencia**

Usando la herencia podemos reutilizar el código de la clase principal. La herencia nos permite definir una clase que hereda todos los métodos y propiedades de la clase principal. La clase padre o superclase o clase base es la clase que proporciona todos los métodos y propiedades. La clase secundaria es la clase que hereda de otra clase principal. Vamos a crear una clase de estudiante heredando de la clase de persona.

In [91]:
class Student(Person):
    pass


s1 = Student('Eyob', 'Yetayeh', 30, 'Finland', 'Helsinki')
s2 = Student('Lidiya', 'Teklemariam', 28, 'Finland', 'Espoo')
print(s1.person_info())
s1.add_skill('JavaScript')
s1.add_skill('React')
s1.add_skill('Python')
print(s1.skills)

print(s2.person_info())
s2.add_skill('Organizing')
s2.add_skill('Marketing')
s2.add_skill('Digital Marketing')
print(s2.skills)

Eyob Yetayeh is 30 years old. He lives in Helsinki, Finland.
['JavaScript', 'React', 'Python']
Lidiya Teklemariam is 28 years old. He lives in Espoo, Finland.
['Organizing', 'Marketing', 'Digital Marketing']


No llamamos al constructor `init()` en la clase secundaria. Si no lo llamamos, aún podemos acceder a todas las propiedades desde el padre. Pero si llamamos al constructor, podemos acceder a las propiedades principales llamando a super .
Podemos agregar un nuevo método al elemento secundario o podemos anular los métodos de la clase principal creando el mismo nombre de método en la clase secundaria. Cuando agregamos la función `init()`, la clase secundaria ya no heredará la función `init()` de los padres.

### **Anulación del método principal**

In [92]:
class Student(Person):
    def __init__ (self, firstname='Asabeneh', lastname='Yetayeh',age=250, country='Finland', city='Helsinki', gender='male'):
        self.gender = gender
        super().__init__(firstname, lastname,age, country, city)
    def person_info(self):
        gender = 'He' if self.gender =='male' else 'She'
        return f'{self.firstname} {self.lastname} is {self.age} years old. {gender} lives in {self.city}, {self.country}.'

s1 = Student('Eyob', 'Yetayeh', 30, 'Finland', 'Helsinki','male')
s2 = Student('Lidiya', 'Teklemariam', 28, 'Finland', 'Espoo', 'female')
print(s1.person_info())
s1.add_skill('JavaScript')
s1.add_skill('React')
s1.add_skill('Python')
print(s1.skills)

print(s2.person_info())
s2.add_skill('Organizing')
s2.add_skill('Marketing')
s2.add_skill('Digital Marketing')
print(s2.skills)

Eyob Yetayeh is 30 years old. He lives in Helsinki, Finland.
['JavaScript', 'React', 'Python']
Lidiya Teklemariam is 28 years old. She lives in Espoo, Finland.
['Organizing', 'Marketing', 'Digital Marketing']


Podemos usar la función integrada `super()` o el nombre del padre `Persona` para heredar automáticamente los métodos y propiedades de su padre. En el ejemplo anterior, anulamos el método principal. El método del niño tiene una característica diferente, puede identificar si el género es masculino o femenino y asignar el pronombre adecuado (Él/Ella).

| **Inicio** | **atrás 12** | **Siguiente 14** |
|----------- |-------------- |---------------|
| [🏠](../../README.md) | [⏪](./12.DataFrames_con_Pandas.ipynb)| [⏩](./14.Scripts_Modulos.ipynb)|